lib/libvfio-user.c

```diff
@@ -927,126 +901,147 @@ device_reset(vfu_ctx_t *vfu_ctx, vfu_reset_type_t reason)
     }
 
     if (vfu_ctx->migration != NULL) {
-        return handle_device_state(vfu_ctx, vfu_ctx->migration,
-                                   VFIO_DEVICE_STATE_V1_RUNNING, false);
+        migr_state_transition(vfu_ctx->migration,
+                              VFIO_USER_DEVICE_STATE_RUNNING);
     }
     return 0;
 }
 
+static uint32_t
+device_feature_flags(uint32_t feature) {
+    switch (feature) {
+        case VFIO_DEVICE_FEATURE_MIGRATION:
+        case VFIO_DEVICE_FEATURE_DMA_LOGGING_REPORT:
+            return VFIO_DEVICE_FEATURE_GET | VFIO_DEVICE_FEATURE_PROBE;
+        case VFIO_DEVICE_FEATURE_MIG_DEVICE_STATE:
+            return VFIO_DEVICE_FEATURE_GET
+                | VFIO_DEVICE_FEATURE_SET
+                | VFIO_DEVICE_FEATURE_PROBE;
+        case VFIO_DEVICE_FEATURE_DMA_LOGGING_START:
+        case VFIO_DEVICE_FEATURE_DMA_LOGGING_STOP:
+            return VFIO_DEVICE_FEATURE_SET | VFIO_DEVICE_FEATURE_PROBE;
+        default:
+            return 0;
+    };
+}
+
 static int
-handle_dirty_pages_get(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
+handle_device_feature(vfu_ctx_t *vfu_ctx, vfu_msg_t *msg)
 {
-    struct vfio_user_dirty_pages *dirty_pages_in;
-    struct vfio_user_dirty_pages *dirty_pages_out;
-    struct vfio_user_bitmap_range *range_in;
-    struct vfio_user_bitmap_range *range_out;
-    size_t argsz;
-    int ret;
+    assert(vfu_ctx != NULL);
+    assert(msg != NULL);
 
+    if (vfu_ctx->migration == NULL ||
```

Comment:
is this right? Shouldn't a client be able to probe for features even if the library user doesn't implement migration?

Reply:
I'm pretty sure the spec says if the probe is unsuccessful (because the feature is unsupported) then we return an error. Since all the features are for migration, it makes sense to have this check here...?

Reply:
We should do this only if we see one of the migration features, not in general. If *no* features are specified we shouldn't say we don't support them!

Reply:
The client has to specify exactly one feature though, and all features are migration features?

Reply:
I'm arguing the test logically belongs in device_feature_flags_supported.